File: app/src/main/java/com/arif/notifforwarder/NotificationForwarderService.java
package com.arif.notifforwarder;

import android.app.Notification;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.util.Log;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public class NotificationForwarderService extends NotificationListenerService {

    private static final String TAG = "NotifForwarder";

    private static final Set<String> IGNORED = new HashSet<>(Arrays.asList(
            "android",
            "com.android.systemui",
            "com.android.settings",
            "com.arif.notifforwarder"
    ));

    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        SharedPreferences prefs = getSharedPreferences("notif_prefs", MODE_PRIVATE);
        if (!prefs.getBoolean("enabled", true)) return;

        final String botToken = prefs.getString("bot_token", "");
        final String chatId   = prefs.getString("chat_id", "");
        final String prefix   = prefs.getString("prefix", "Notifikasi dari HP:");

        if (botToken == null || botToken.isEmpty()) return;
        if (chatId   == null || chatId.isEmpty())   return;
        if (IGNORED.contains(sbn.getPackageName()))  return;
        if (sbn.isOngoing()) return;

        // Filter app
        if (prefs.getBoolean("filter_enabled", false)) {
            Set<String> allowed = prefs.getStringSet("allowed_packages", new HashSet<>());
            if (allowed == null || allowed.isEmpty()) return;
            if (!allowed.contains(sbn.getPackageName())) return;
        }

        Bundle extras  = sbn.getNotification().extras;
        String title   = extras.getString(Notification.EXTRA_TITLE, "");
        CharSequence cs = extras.getCharSequence(Notification.EXTRA_BIG_TEXT);
        if (cs == null) cs = extras.getCharSequence(Notification.EXTRA_TEXT);
        final String body = (cs != null) ? cs.toString() : "";

        if (title.isEmpty() && body.isEmpty()) return;

        String appName;
        try {
            appName = getPackageManager()
                    .getApplicationLabel(getPackageManager()
                            .getApplicationInfo(sbn.getPackageName(), 0))
                    .toString();
        } catch (Exception e) {
            appName = sbn.getPackageName();
        }

        final String fTitle   = title;
        final String fApp     = appName;
        final String fToken   = botToken;
        final String fChat    = chatId;
        final String time     = new SimpleDateFormat("HH:mm", Locale.getDefault()).format(new Date());

        final String message = prefix + "\n\n"
                + "*App:* "    + TelegramSender.escape(appName) + "\n"
                + (title.isEmpty() ? "" : "*Judul:* " + TelegramSender.escape(title) + "\n")
                + (body.isEmpty()  ? "" : "*Pesan:* " + TelegramSender.escape(body)  + "\n")
                + "*Waktu:* " + time;

        new Thread(new Runnable() {
            @Override
            public void run() {
                NotifQueueDbHelper db = NotifQueueDbHelper.getInstance(getApplicationContext());
                db.insert(message, fToken, fChat);

                if (NetworkChangeReceiver.isConnected(getApplicationContext())) {
                    List<NotifQueueDbHelper.QueueItem> all = db.getAll();
                    for (int i = all.size() - 1; i >= 0; i--) {
                        NotifQueueDbHelper.QueueItem q = all.get(i);
                        if (q.message.equals(message) && q.botToken.equals(fToken)) {
                            boolean sent = TelegramSender.send(fToken, fChat, message);
                            if (sent) {
                                db.deleteById(q.id);
                                Log.d(TAG, "Terkirim: " + fApp);
                            } else {
                                Log.w(TAG, "Gagal, masuk antrian: " + fApp);
                            }
                            break;
                        }
                    }
                } else {
                    Log.w(TAG, "Offline, masuk antrian: " + fApp);
                }
            }
        }).start();
    }
}

File: app/src/main/java/com/notifforwarder/app/NotificationForwarderService.kt
package com.notifforwarder.app

import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.util.Log
import kotlinx.coroutines.*
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.text.SimpleDateFormat
import java.util.*

class NotificationForwarderService : NotificationListenerService() {

    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val ignoredPackages = setOf(
        "android",
        "com.android.systemui",
        "com.android.settings",
        "com.notifforwarder.app"
    )

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        val prefs = getSharedPreferences("notif_prefs", MODE_PRIVATE)

        if (!prefs.getBoolean("enabled", true)) return

        val botToken = prefs.getString("bot_token", "") ?: return
        val chatId   = prefs.getString("chat_id", "")   ?: return
        val prefix   = prefs.getString("prefix", "📱 Notifikasi dari HP:") ?: ""

        if (botToken.isEmpty() || chatId.isEmpty()) return
        if (sbn.packageName in ignoredPackages) return
        if (sbn.isOngoing) return

        // ── CEK FILTER APP ────────────────────────────────────────────────
        val filterEnabled = prefs.getBoolean("filter_enabled", false)
        if (filterEnabled) {
            val allowed = prefs.getStringSet("allowed_packages", emptySet()) ?: emptySet()
            if (allowed.isEmpty() || sbn.packageName !in allowed) return
        }
        // ──────────────────────────────────────────────────────────────────

        val extras  = sbn.notification.extras
        val title   = extras.getString("android.title") ?: ""
        val text    = extras.getCharSequence("android.text")?.toString() ?: ""
        val bigText = extras.getCharSequence("android.bigText")?.toString() ?: ""
        val body    = if (bigText.isNotEmpty()) bigText else text

        if (title.isEmpty() && body.isEmpty()) return

        val appName = try {
            val info = packageManager.getApplicationInfo(sbn.packageName, 0)
            packageManager.getApplicationLabel(info).toString()
        } catch (e: Exception) { sbn.packageName }

        val time = SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date())

        val message = buildString {
            appendLine(prefix)
            appendLine()
            appendLine("📦 *App:* ${escapeMarkdown(appName)}")
            if (title.isNotEmpty()) appendLine("📌 *Judul:* ${escapeMarkdown(title)}")
            if (body.isNotEmpty())  appendLine("💬 *Pesan:* ${escapeMarkdown(body)}")
            appendLine("🕐 *Waktu:* $time")
        }

        serviceScope.launch {
            val db  = AppDatabase.getInstance(applicationContext)
            val dao = db.notificationDao()

            // 1. Simpan ke antrian lokal DULU
            val queued = QueuedNotification(
                message  = message,
                botToken = botToken,
                chatId   = chatId
            )
            dao.insert(queued)

            // 2. Coba kirim langsung
            val sent = sendToTelegram(botToken, chatId, message)

            if (sent) {
                // Berhasil → hapus dari antrian
                val justSent = dao.getAll().lastOrNull {
                    it.message == message && it.botToken == botToken
                }
                if (justSent != null) dao.delete(justSent)
                Log.d("NotifForwarder", "Terkirim langsung: $appName")
            } else {
                // Gagal (tidak ada internet) → jadwalkan WorkManager
                Log.w("NotifForwarder", "Tidak ada internet, masuk antrian: $appName")
                SendQueueWorker.schedule(applicationContext)
            }
        }
    }

    private fun sendToTelegram(botToken: String, chatId: String, message: String): Boolean {
        return try {
            val url  = URL("https://api.telegram.org/bot$botToken/sendMessage")
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            conn.doOutput       = true
            conn.connectTimeout = 8000
            conn.readTimeout    = 8000

            val params = "chat_id=${URLEncoder.encode(chatId, "UTF-8")}" +
                    "&text=${URLEncoder.encode(message, "UTF-8")}" +
                    "&parse_mode=Markdown"

            conn.outputStream.use { it.write(params.toByteArray()) }
            val code = conn.responseCode
            conn.disconnect()
            code == 200
        } catch (e: Exception) {
            false
        }
    }

    private fun escapeMarkdown(text: String): String {
        return text.replace("_", "\\_")
            .replace("*", "\\*")
            .replace("[", "\\[")
            .replace("`", "\\`")
    }

    override fun onDestroy() {
        super.onDestroy()
        serviceScope.cancel()
    }
}
